File: asherah-java/java/src/main/java/com/godaddy/asherah/appencryption/SessionBytesImpl.java
package com.godaddy.asherah.appencryption;

import com.godaddy.asherah.jni.AsherahSession;
import org.json.JSONObject;

import java.nio.charset.StandardCharsets;

/**
 * Session implementation for byte[] payloads.
 * Wraps the native AsherahSession. Compatible with the canonical godaddy/asherah SDK.
 *
 * @param <D> the data row record type (byte[] or JSONObject)
 */
class SessionBytesImpl<D> implements Session<byte[], D> {

    private final AsherahSession inner;
    private final boolean drrIsJson;

    SessionBytesImpl(final AsherahSession inner, final boolean drrIsJson) {
        this.inner = inner;
        this.drrIsJson = drrIsJson;
    }

    @Override
    @SuppressWarnings("unchecked")
    public D encrypt(final byte[] payload) {
        final String drrJson = inner.encryptToJson(payload);
        if (drrIsJson) {
            return (D) new JSONObject(drrJson);
        }
        return (D) drrJson.getBytes(StandardCharsets.UTF_8);
    }

    @Override
    public byte[] decrypt(final D dataRowRecord) {
        final String drrJson;
        if (dataRowRecord instanceof JSONObject) {
            drrJson = ((JSONObject) dataRowRecord).toString();
        } else {
            drrJson = new String((byte[]) dataRowRecord, StandardCharsets.UTF_8);
        }
        return inner.decryptFromJson(drrJson);
    }

    @Override
    public void close() {
        inner.close();
    }
}

File: asherah-dotnet/AsherahDotNet/AppEncryption/SessionBytesImpl.cs
using System.Text;
using Newtonsoft.Json.Linq;

namespace GoDaddy.Asherah.AppEncryption;

internal class SessionBytesImpl<TD> : Session<byte[], TD>
{
    private readonly AsherahSession _inner;
    private readonly bool _drrIsJson;

    internal SessionBytesImpl(AsherahSession inner, bool drrIsJson)
    {
        _inner = inner;
        _drrIsJson = drrIsJson;
    }

    public override TD Encrypt(byte[] payload)
    {
        var drrJson = Encoding.UTF8.GetString(_inner.EncryptBytes(payload));
        if (_drrIsJson)
            return (TD)(object)JObject.Parse(drrJson);
        return (TD)(object)Encoding.UTF8.GetBytes(drrJson);
    }

    public override byte[] Decrypt(TD dataRowRecord)
    {
        string drrJson;
        if (dataRowRecord is JObject jobj)
            drrJson = jobj.ToString(Newtonsoft.Json.Formatting.None);
        else
            drrJson = Encoding.UTF8.GetString((byte[])(object)dataRowRecord!);

        return _inner.DecryptBytes(Encoding.UTF8.GetBytes(drrJson));
    }

    public override void Dispose() => _inner.Dispose();
}
